### PR TITLE
add python conveniences to mathtext

### PR DESCRIPTION
**Pull request by IuryPiva**
- [x] issues: fixes #11395
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

Implemented suggestions from #11395

a user could use:

- `p.xaxis_label = r"$\sin(x)$"`
- `p.xaxis_label = MathText(text=r"\sin(x)")`
- `p.xaxis_label = MathText(r"\sin(x)")`
any of the above example will result in the same `sin(x)` rendered as **LaTeX**

The examples below will result in `$sin(x)$` rendered as **LaTeX**
- `p.xaxis_label = MathText(r"$\sin(x)$")`
- `p.xaxis_label = MathText(text=r"$\sin(x)$")`

The examples below will result in `$sin(x)$` rendered as **plain text**
- `p.xaxis_label = PlainText(r"$\sin(x)$")`
- `p.xaxis_label = PlainText(text=r"$\sin(x)$")`

### CODE REVIEW

File: bokeh/models/axes.py [hunk 1]
```diff
@@ -124,10 +124,22 @@ class Axis(GuideRenderer):
...
+    def axis_label(self) -> Nullable(Either(String, Instance(MathText))):
+        return self._axis_label
+
+    @axis_label.setter
+    def axis_label(self, value: Either(String, Instance(MathText))) -> None:
+        if (isinstance(value, str)):
+            if(value.startswith("$") and value.endswith("$")):
+                self._axis_label = MathText(text=value)
+        else:
+            self._axis_label = value
```
Comment (bryevdv) [inline bokeh/models/axes.py:141]: This is not the right way to go about this, for a few reasons. We should use a `.accepts` clause on this. Ideally we could define a new string/regex property `MathString` type that matches `$......$` strings, then 
```
axis_label = Nullable(Either(String, Instance(MathText)), ...).accepts(MathString, lambda val: MathText(text=val))
```

Comment (bryevdv) [reply bokeh/models/axes.py:141]: a `MathString` might look like:
```
class MathString(String):
    def validate(self, value, detail=True):
        super().validate(value, detail)

        if  len(value) > = 2 and value[0] == value[-1] == "$":
            return

        msg = "" if not detail else f"expected a math string"
        raise ValueError(msg)
```

Comment (IuryPiva) [reply bokeh/models/axes.py:141]: > axis_label = Nullable(...).accepts(MathString, lambda val: MathText(text=val[1:-1]))

wouldn't this imply that we only accept math strings? we agreed on accepting all of [these formats](https://github.com/bokeh/bokeh/discussions/11395#discussioncomment-953157)



I will add the `PlainText` model to allow strings without interpretation.

Comment (bryevdv) [reply bokeh/models/axes.py:141]: @IuryPiva No, the `accepts` conversion function is only used if the value type-validates against the `accepts` type.  In this case `MathText` will only validate strings like `$.....$` I think we may not need `PlainText` it's possible the existing `String` will suffice to mark plain text

Comment (IuryPiva) [reply bokeh/models/axes.py:141]: Alright, thanks Bryan

Comment (IuryPiva) [reply bokeh/models/axes.py:141]: Can you explain what corner cases are you thinking of when you mention that `$` might be in the center of the string?

**Comment by IuryPiva** [comment]
@bryevdv it is not converting to MathText with the `.accepts` clause. Do you see what I am doing wrong here?

**Comment by bryevdv** [comment]
@IuryPiva it looks like we will need to do a little more thinking. The `accepts` alternatives are only *fallbacks*, used if the main property validation fails:

https://github.com/bokeh/bokeh/blob/9bfa131130da3a6905654b1f9f4388c87f1f91b0/bokeh/core/property/bases.py#L344-L353

I had thought they applied earlier and could supersede the main property type, but that's not the case.

@mattpap what are your thoughts here? We could add a flag to `accepts` to enable checking of the accepts before the main property, or we could define new `preaccepts` (or whatever) to signal the same. Additionally, I am not sure it's clearly documented anywhere that `accepts` can only function when validation is on, but we should make sure that is clearly stated some place.

File: bokeh/models/axes.py (hunk 1 above)
Comment (bryevdv) [reply bokeh/models/axes.py]: I just mean it's always good to consider explicit what can, should, and does happen in corner cases. In this situation, things like `"$...$..$"` or `"$...\$....$" or `"$$...$"` etc.  We should make sure there are tests that cover cases like this *now*, so that if the implementation ever changes in the future, they will already be covered.

File: bokehjs/src/lib/core/types.ts [hunk 2]
```diff
@@ -1,3 +1,6 @@
+import {MathText} from "models/math_text"
+import {PlainText} from "models/plain_text"
```
[the remark this answers is not in the document]
Comment (IuryPiva) [reply bokehjs/src/lib/core/types.ts:2]: Do you think I should move `TextLike` type out of here?

File: bokeh/core/property/text_like.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+__all__ = (
+    'TextLike',
+)
+
+#-----------------------------------------------------------------------------
+# General API
+#-----------------------------------------------------------------------------
+
+TextLike = Either(MathString, Instance("bokeh.models.math_text.MathText"), Instance("bokeh.models.plain_text.PlainText"))
```
[the remark this answers is not in the document]
Comment (IuryPiva) [reply bokeh/core/property/text_like.py]: I agree, the new `models/text.py` seems like a good place to it.

File: sphinx/source/docs/user_guide/styling.rst [hunk 4]
```diff
@@ -1117,7 +1117,7 @@ only with :ref:`axis labels <userguide_styling_axes_labels>`. Supporting LaTeX
 notations in more elements is planned for future releases.
 
 Bokeh uses the MathJax_ library to handle LaTeX markup. To use LaTeX markup
-instead of plain text, use a :class:`~bokeh.models.math_text.MathText` object:
+instead of plain text, use a :class:`~bokeh.models.text.MathText` object:
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/styling.rst:1120]: The split of `bokeh.models` into sub-modules is purely for maintenance considerations. Users should always (and only) be instructed to import from top-level `bokeh.models`

Comment (bryevdv) [reply sphinx/source/docs/user_guide/styling.rst:1120]: Tho I guess refs like this might need to stay the way they are currently, until we can completely re-arrange the reference guide. Disregard the previous comment, for this PR.

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
LGTM @mattpap I'll plan to merge tomorrow unless you have more comments (or merge first)
